### PR TITLE
Fix infinite loop in `_validate_not_a_forked_repo()`

### LINKED ISSUE #61755: HTTP Error 403 for torch.hub ResNet
## 🐛 Bug

`torch.hub` pretrained resnet doesn't work.

## To Reproduce

On colab:

```py
import torch

model = torch.hub.load('pytorch/vision:v0.9.0', 'resnet18', pretrained=True)
```

Error:

```py
/usr/lib/python3.7/urllib/request.py in http_error_default(self, req, fp, code, msg, hdrs)
    647 class HTTPDefaultErrorHandler(BaseHandler):
    648     def http_error_default(self, req, fp, code, msg, hdrs):
--> 649         raise HTTPError(req.full_url, code, msg, hdrs, fp)
    650 
    651 class HTTPRedirectHandler(BaseHandler):

HTTPError: HTTP Error 403: rate limit exceeded
```

This is identical to the code snippet on the [website](https://pytorch.org/hub/pytorch_vision_resnet/)

## Expected behavior

The code should run.

## Environment

Google colab.

cc @nairbv

### PR DESCRIPTION
**Pull request by malfet**
Increase `page_idx` in the loop rather than outside of it
Break from the loop when receive empty response as it means there are no more items to fetch via pagination request

Also, add options to use provided github token (via `GITHUB_TOKEN` environment variable) 
   
Fixes failure with "Rate Limit Exceeded" when doing something like `torch.hub.list("pytorch/test-infra:dsf")`

Fixes #61755

### CODE REVIEW

File: torch/hub.py [hunk 1]
```diff
@@ -113,28 +114,37 @@ def _parse_repo_info(github):
...
-
-        page += 1
+            page += 1
+            url = f'{url_prefix}?per_page=100&page={page}'
+            response = json.loads(_read_url(Request(url, headers=headers)))
+            # Empty response means no more data to process
+            if not response:
+                break
+            for br in response:
+                if br['name'] == branch or br['commit']['sha'].startswith(branch):
```
Comment (mthrok) [inline torch/hub.py:140]: Not: Wonder if we should check the existence of the key. It's very unlikely to happen but we could give better error message.

**Review by mthrok — APPROVED** [review]
LGTM

File: torch/hub.py (hunk 1 above)
Comment (malfet) [reply torch/hub.py:140]: Good point let me address that

**Comment by malfet** [comment]
Moved `skipped_validation` to #62139

**Pull request closed without merging**